internal/component/kafka/consumer.go

```diff
@@ -35,28 +36,110 @@ type consumer struct {
+				}
+			} else {
+				err := consumer.doCallback(session, message)
+				if err != nil {
+					consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				}
+			}
+		}
+	}
+	if isBulkSubscribe {
 		if consumer.k.consumeRetryEnabled {
 			if err := retry.NotifyRecover(func() error {
-				return consumer.doCallback(session, message)
+				return consumer.doBulkCallback(session, messages, claim.Topic())
 			}, b, func(err error, d time.Duration) {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v. Retrying...", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Warnf("Error processing Kafka bulk messages: %s. Error: %v. Retrying...", claim.Topic(), err)
 			}, func() {
-				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
+				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s", claim.Topic())
 			}); err != nil {
-				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s. Error: %v.", claim.Topic(), err)
 			}
 		} else {
-			err := consumer.doCallback(session, message)
+			err := consumer.doBulkCallback(session, messages, claim.Topic())
 			if err != nil {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+				consumer.k.logger.Errorf("Error processing Kafka message: %s. Error: %v.", claim.Topic(), err)
 			}
 		}
 	}
-
 	return nil
 }
 
+func (consumer *consumer) doBulkCallback(session sarama.ConsumerGroupSession, messages []*sarama.ConsumerMessage, topic string) error {
+	consumer.k.logger.Debugf("Processing Kafka bulk message: %s", topic)
+	handler, err := consumer.k.GetTopicBulkHandler(topic)
+	if err != nil {
+		return err
+	}
+	messageValues := make([]KafkaBulkMessageEntry, (len(messages)))
+
+	for i, message := range messages {
+		if message != nil {
+			metadata := make(map[string]string, len(message.Headers))
+			if message.Headers != nil {
+				for _, t := range message.Headers {
+					metadata[string(t.Key)] = string(t.Value)
+				}
+			}
+			entryID, entryIDErr := uuid.NewRandom()
+			if entryIDErr != nil {
+				consumer.k.logger.Errorf("Failed to generate entryID for sending message for bulk subscribe event on topic: %s. Error: %v.", topic, err)
```

Comment:
You cannot really proceed if you got an error here. I suggest we skip this message and mark it as errored in the return automatically.

Reply:
I could think of 4 options here:
1. return from func doBulkCallback() with an error as system doesn't seem to have enough entropy (if I understand correctly). Not opting for it as found Option 4 to be better.
2. don't pass it on this message and mark with an error in the return automatically - Right now, I depend on runtime, to process and pass back responses to me. And, I mark the messages as per theses reponses. But, to take care of this message on the kafka implementation side, I would have to create a new response array here itself and track of the indexes which I didn't send, in some way. So, want to avoid this.
Infact, if an entryID is passed as blank, runtime is expected to pass back an error pertaining to this message at the same index in return reponse array. ()
3. Pass a blank entryID here and rely on runtime to provide back an error for this message.
4. use index as entryID here.

I am opting for 4th option. Please let me know if that works.